src/reduction/mod.rs

```diff
@@ -1,247 +1,115 @@
-    if grayscale_reduction && reduced.ihdr.color_type.is_rgb() {
-        if let Some(r) = reduce_rgb_to_grayscale(&reduced) {
-            reduced = Cow::Owned(r);
+    // This is just removal of bytes and does not need to be evaluated
+    if opts.color_type_reduction && !deadline.passed() {
+        if let Some(reduced) = reduce_rgb_to_grayscale(&png) {
+            png = Arc::new(reduced);
+            reduction_occurred = true;
         }
     }
 
-    // Attempt grayscale alpha reduction before palette, as grayscale will typically be smaller than indexed
-    if reduced.ihdr.color_type == ColorType::GrayscaleAlpha {
-        if let Some(r) = reduced_alpha_channel(&reduced, optimize_alpha) {
-            reduced = Cow::Owned(r);
+    // Now retain the current png for the evaluator baseline
+    // It will only be entered into the evaluator if there are also others to evaluate
+    let mut baseline = png.clone();
+
+    // Attempt alpha removal
+    if opts.color_type_reduction && !deadline.passed() {
+        if let Some(reduced) = reduced_alpha_channel(&png, opts.optimize_alpha) {
+            png = Arc::new(reduced);
+            // If the reduction requires a tRNS chunk, enter this into the evaluator
+            // Otherwise it is just removal of bytes and should become the baseline
+            if png.ihdr.color_type.has_trns() {
+                eval.try_image(png.clone());
+                evaluation_added = true;
+            } else {
+                baseline = png.clone();
+                reduction_occurred = true;
+            }
         }
     }
 
-    // Attempt to reduce to palette, if not already a single channel
-    if reduced.channels_per_pixel() != 1 {
-        if let Some(r) = reduce_to_palette(&reduced) {
-            reduced = Cow::Owned(r);
-
-            // Make sure that palette gets sorted. Ideally, this should be done within reduce_to_palette.
-            if let Some(r) = reduced_palette(&reduced, optimize_alpha) {
-                reduced = Cow::Owned(r);
-            }
+    // Attempt to reduce the palette size
+    if opts.palette_reduction && !deadline.passed() {
+        if let Some(reduced) = reduced_palette(&png, opts.optimize_alpha) {
+            png = Arc::new(reduced);
+            eval.try_image(png.clone());
+            evaluation_added = true;
         }
     }
 
-    // Attempt RGBA alpha reduction after palette, so it can be skipped if palette was successful
-    if reduced.ihdr.color_type == ColorType::RGBA {
-        if let Some(r) = reduced_alpha_channel(&reduced, optimize_alpha) {
-            reduced = Cow::Owned(r);
+    // Attempt to reduce to palette
+    if opts.color_type_reduction && !deadline.passed() {
+        if let Some(reduced) = reduce_to_palette(&png) {
+            png = Arc::new(reduced);
+            // Make sure the palette gets sorted (ideally, this should be done within reduce_to_palette)
+            if let Some(reduced) = reduced_palette(&png, opts.optimize_alpha) {
```

Comment:
can `reduce_to_palette` be moved before line 80 to try `reduced_palette` once?

Reply:
The two calls to `reduced_palette` are mutually exclusive: The earlier one will run if the image is already indexed (the function will simply return if it isn't), while this one will run if we've just converted to indexed.